### ISSUE TITLE
Create default base view w auth, object permissions

### ISSUE THREAD

**Issue by bengolder** [codeforamerica/intake#658]
### What is the goal of this task?

- To implement access controls in a consistent, centralized manner that is more maintainable for future development.

This would involve writing the following pieces of code:

#### Base view and functions for easier permissions

For ease of future implementation and maintenance, we would a base view with hooks for permissions checking, and functions that can be used outside of views or in views that deviate from the pattern.

**Comment by bengolder** [comment]
### Update: There may be a more efficient and maintainable approach.

Talked with @joduinn, @glassresistor and both advocated for an approach that would log at the database layer rather than the application layer.

Advantages of logging at the DB layer:
- no application code to maintain
- adds logging coverage for any request handling, whether performed by the Django framework's built-in admin or by any 3rd party libraries.
- reduces risk of human error or bad intention, because it would only require access to the database configuration rather than writing compliant source code
- would likely be a lower performance drag than logging from the application layer
- closer to the source of truth

Possible blockers to logging at the DB layer:
- it may not be feasible to identify the user or individual for each event.
- it may not be feasible to identify the outcome (failed or succeeded) of a given user action.
- it may not be feasible to customize the logging from the RDS database to accommodate any of this

Next Steps:
- Research the feasibility of logging at the DB layer from RDS
    - How does it work?
    - How can we configure it, if at all?
    - Can logs be directed to Cloudwatch?
    - Can we identify users in the context of any DB query?
    - Can we identify the success or failure of a an attempt to access a sensitive resource?

In summary, if we can log at the database layer, it would be far cheaper and more maintainable. But we don't know if it's possible. If not, we should continue with plan of logging from application layer.

**Comment by glassresistor** [comment]
I went ahead and did some research on the feasibility of logging at the database layer.  Postgres is not configurable to log the results of queries and therefore is not suitable for our needs as the main audit source.  We should still make sure we log every query for many other reasons.

Given that direct database logging of the results is not feasible the next best option is to have the application log the necessary information on each request.  This will be much cheaper than having the application log each query and its result as there are much fewer requests than queries and should be able to connect this with the request logging thats already happening.

Doing it on the request ends up being more convenient as well because the request and response object should be readily available at this point meaning we can log the request and the context we send to the template(so all the data we should the user).

**Comment by jennifermarie** [comment]
## Key recommendations from security review
- Have consistent access control
- For any policy, log your compliance with your own access control
- Read your logs
- Protect the logs from editing, and log read access to the logs themselves.

**Comment by jennifermarie** [comment]
We received a lot of suggestions for alternative ways to implement these features. We took some time out to spike out prototype code to help us assess the pros and cons of each.


### Suggestion 1: Log at the database layer because low cost & fully automated

Pros:
- fully automated, doesn't depend on dev compliance
- includes everything we would need for forensic & more

Cons:
- too much information, would end up logging a lot of PII, creating an additional privacy risk.
- If we logged indiscriminately, it would add a performance.
- difficult to configure the logs to exclude and filter out things we don't want.
- Would require dev work to include context of the request and authenticated user
- This doesn't include access control checks


### Suggestion 2: Use Django's built in `signals` API
Pros:
- Django has a "signals" feature that fires when certain actions are performed on any model
- This is very similar to an event pub/sub, you make listener functions and listen for events on certain models.

Cons:
- This is specifically recommended against by "Two Scoops of Django" due to spaghetti code and for commonly being misinterpreted as asynchronous
- It also does not cover reads
- It does not cover any bulk actions (bulk update or create, which we use)
- This does not include access control
- We tried it and found ourselves in an infinite-loop situation

### Suggestion 3: DEBUG log -- log everything at the app layer
Pros:
- Low dev cost; just log everything to a crazy degree

Cons:
- Possible high performance cost
- It is theoretically quite possible to spend some dev time filtering what is logged, can be scoped


### Suggestion 4: Database triggers (access event listeners on certain tables that create log rows in the database)
Pros:
- (presumably) Low-maintenance once implemented
- No need to refactor existing code

Cons:
- Would create a new layer of code to maintain DB-level scripts
- Doesn't necessarily come with the level of logging that others have
- Would need request context (this is possible, not sure of costs)


Following consideration of these alternative options, we have decided to proceed with the middleware solution described above and in issues #778, #779, and #800.

**Comment by bengolder** [comment]
Update: we separated logging from access control. This task remains focused on building consistent access controls, while #779 covers the task of logging crud events from our app and the django admin.
